src/axislike.rs

```diff
@@ -797,53 +781,58 @@ impl std::hash::Hash for DeadZoneShape {
 
 impl DeadZoneShape {
     /// Returns whether the (x, y) input is outside the deadzone.
-    pub fn input_outside_deadzone(&self, x: f32, y: f32) -> bool {
+    pub fn deadzone_input_value(&self, x: f32, y: f32) -> Option<DualAxisData> {
+        let value = Vec2::new(x, y);
+
         match self {
             DeadZoneShape::Cross {
-                rect_1_width,
-                rect_1_height,
-                rect_2_width,
-                rect_2_height,
-            } => self.outside_cross(
-                x,
-                y,
-                *rect_1_width,
-                *rect_1_height,
-                *rect_2_width,
-                *rect_2_height,
-            ),
-            DeadZoneShape::Rect { width, height } => self.outside_rectangle(x, y, *width, *height),
+                horizontal_width,
+                vertical_width,
+            } => self.cross_deadzone_value(value, *horizontal_width, *vertical_width),
             DeadZoneShape::Ellipse { radius_x, radius_y } => {
-                self.outside_ellipse(x, y, *radius_x, *radius_y)
+                self.ellipse_deadzone_value(value, *radius_x, *radius_y)
             }
         }
     }
 
-    /// Returns whether the (x, y) input is outside a cross.
-    fn outside_cross(
+    /// Computes the input value based on the cross deadzone.
+    fn cross_deadzone_value(
         &self,
-        x: f32,
-        y: f32,
-        rect_1_width: f32,
-        rect_1_height: f32,
-        rect_2_width: f32,
-        rect_2_height: f32,
-    ) -> bool {
-        self.outside_rectangle(x, y, rect_1_width, rect_1_height)
-            && self.outside_rectangle(x, y, rect_2_width, rect_2_height)
-    }
-
-    /// Returns whether the (x, y) input is outside a rectangle.
-    fn outside_rectangle(&self, x: f32, y: f32, width: f32, height: f32) -> bool {
-        x >= width || x <= -width || y >= height || y <= -height
-    }
-
-    /// Returns whether the (x, y) input is outside an ellipse.
-    fn outside_ellipse(&self, x: f32, y: f32, radius_x: f32, radius_y: f32) -> bool {
-        if radius_x == 0.0 || radius_y == 0.0 {
-            return true;
+        value: Vec2,
+        horizontal_width: f32,
+        vertical_width: f32,
+    ) -> Option<DualAxisData> {
+        let new_x = f32::from(value.x.abs() > vertical_width) * value.x;
+        let new_y = f32::from(value.y.abs() > horizontal_width) * value.y;
+        let new_value = Vec2::new(new_x, new_y);
+
+        if new_value == Vec2::ZERO {
+            None
+        } else {
+            let scaled_value =
+                Self::scale_value(new_value, Vec2::new(vertical_width, horizontal_width));
+            Some(DualAxisData::from_xy(scaled_value))
         }
+    }
+
+    /// Computes the input value based on the ellipse deadzone.
+    fn ellipse_deadzone_value(
+        &self,
+        value: Vec2,
+        radius_x: f32,
+        radius_y: f32,
+    ) -> Option<DualAxisData> {
+        let radius_x = radius_x.max(0.0001);
```

Comment:
I just picked a random small number to make sure that we aren't dividing by 0.0. Is there is more standard number that is typically used in this situation, like epsilon? I don't think I can return `None` or `Some` when the deadzone size is 0.0 as then all values are ignored or read.

Reply:
f32::EPSILON is probably the most "standard", but just extracting it into a constant called `EPSILON` is probably clear enough.